File: services/nuldoc/nuldoc-src/generators/tag.ts
import TagPage from "../pages/TagPage.ts";
import { Config, getTagLabel } from "../config.ts";
import { Page } from "../page.ts";
import { TaggedPage } from "./tagged_page.ts";

export interface TagPage extends Page {
  tagSlug: string;
  tagLabel: string;
  numOfPosts: number;
  numOfSlides: number;
}

export async function generateTagPage(
  tagSlug: string,
  pages: TaggedPage[],
  site: "blog" | "slides",
  config: Config,
): Promise<TagPage> {
  const html = await TagPage(tagSlug, pages, site, config);

  return {
    root: html,
    renderer: "html",
    site,
    destFilePath: `/tags/${tagSlug}/index.html`,
    href: `/tags/${tagSlug}/`,
    tagSlug: tagSlug,
    tagLabel: getTagLabel(config, tagSlug),
    numOfPosts: pages.filter((p) => !("event" in p)).length,
    numOfSlides: pages.filter((p) => "event" in p).length,
  };
}

File: services/nuldoc/lib/nuldoc/generators/tag.rb
module Nuldoc
  TagPageData = Data.define(:root, :renderer, :site, :dest_file_path, :href,
                            :tag_slug, :tag_label, :num_of_posts, :num_of_slides)

  module Generators
    class Tag
      def initialize(tag_slug, pages, site, config)
        @tag_slug = tag_slug
        @pages = pages
        @site = site
        @config = config
      end

      def generate
        html = Pages::TagPage.new(tag_slug: @tag_slug, pages: @pages, site: @site, config: @config).render

        TagPageData.new(
          root: html,
          renderer: :html,
          site: @site,
          dest_file_path: "/tags/#{@tag_slug}/index.html",
          href: "/tags/#{@tag_slug}/",
          tag_slug: @tag_slug,
          tag_label: @config.tag_label(@tag_slug),
          num_of_posts: @pages.count { |p| !p.respond_to?(:event) },
          num_of_slides: @pages.count { |p| p.respond_to?(:event) }
        )
      end
    end
  end
end
